Wrappers/Python/cil/optimisation/operators/BlockOperator.py

```diff
@@ -135,26 +136,43 @@ def get_item(self, row, col):
         index = row*self.shape[1]+col
         return self.operators[index]
     
-    def norm(self, **kwargs):
-        '''Returns the norm of the BlockOperator
-
-        if the operator in the block do not have method norm defined, i.e. they are SIRF
-        AcquisitionModel's we use PowerMethod if applicable, otherwise we raise an Error
+    def norm(self):
+        '''Returns the square root of the sum of the norms of the individual operators in the BlockOperators 
+        '''
+        return numpy.sqrt(numpy.sum(numpy.array(self.norms())**2))    
+    
+    def norms(self, ):
+        '''Returns a list of the individual norms of the Operators in the BlockOperator
         '''
-        norm = []
+        norms= []
         for op in self.operators:
-            if hasattr(op, 'norm'):
-                norm.append(op.norm(**kwargs) ** 2.)
-            else:
-                # use Power method
-                if op.is_linear():
-                    norm.append(
-                            LinearOperator.PowerMethod(op, 20)[0]
-                            )
-                else:
-                    raise TypeError('Operator {} does not have a norm method and is not linear'.format(op))
-        return numpy.sqrt(sum(norm))    
+            try:
+                norms.append(op.norm())
+            except:
+                    raise TypeError('Operator {} does not have a norm method'.format(op))
+        return norms
     
+    def set_norms(self, norms):
+        '''Uses the set_norm() function in Operator to set the norms of the operators in the BlockOperator from a list of custom values. 
+
+
+        '''
+        if len(norms)==len(self.operators):
+            if all(isinstance(i, Number) or i is None for i in norms):
+                if all(  k is None or k>=0 for k in norms ):
+                    pass
+                else:
+                    raise ValueError("Each number in the list should be positive")
+            else: 
+                raise ValueError("Each element in the list of norms should be a number or None")
+        else:
+            raise ValueError("The length of the list of norms should be equal to the number of operators in the BlockOperator")
+        
+        for j,value in enumerate(norms):
+            self.operators[j].set_norm(value)
```

Comment:
Currently it iterates through the list to see if the values are all `Number` or `None` and then again to check they are positive. Then again to get the value and pass it to `set_norm`.

I think it's enough to test each value is positive (and real?) and raise an error if it's not.

This error should probably be raised by a setter for `set_norm` at the operator level. Maybe something like
```
try:
    if  value is not None and value is <0
        raise ValueError("Norm must be a positive real value or None, got X")
except TypeError:
    raise TypeError("Norm must be a positive real value or None, got X of type Y")
```

Which would simplify it to a single line here that just pulls out the values and trusts the `set_norm` to test them.

Reply:
Thanks Gemma

Reply:
Added some tests to set_norm in the operator